[SPARK-27774][CORE][MLLIB] Avoid hardcoded configs

wenxuanguan opened the pull request:
## What changes were proposed in this pull request?

avoid hardcoded configs in `SparkConf` and `SparkSubmit` and test

## How was this patch tested?

N/A

wenxuanguan committed:
avoid hardcoded config

wenxuanguan committed:
fix kafka sink semantics in document

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

dongjoon-hyun commented on docs/structured-streaming-kafka-integration.md:
```diff
@@ -441,7 +441,7 @@ Apache Kafka only supports at least once write semantics. Consequently, when wri
 to retry a message that was not acknowledged by a Broker, even though that Broker received and wrote the message record.
 Structured Streaming cannot prevent such duplicates from occurring due to these Kafka write semantics. However,
-if writing the query is successful, then you can assume that the query output was written at least once. A possible
+if writing the query is successful, then you can assume that the query output was written exactly once. A possible
```
Hi, @wenxuanguan . This looks wrong in this context. Could you explain your thought?

wenxuanguan replied:
Thanks for your reply.
I thought this describes the situation that query writes to kafka successfully and no records are duplicated. How about change to `So if writing the query is successful, then you can assume that the query output was written at least once`, which will not be confused by `However`

dongjoon-hyun replied:
The existing one looks okay to me. Let's remove this doc change from this PR.

wenxuanguan replied:
@dongjoon-hyun Thanks, done

dongjoon-hyun reviewed (changes requested):
Please revert the doc change and remove `[DOC]` and `fix kafka sink write semantics in document ` from the title.

BTW, @wenxuanguan . We usually don't put different stuffs into a single PR. Single PR had better have a single issue. Please make separate PRs for the orthogonal issues next time. Thanks.

wenxuanguan committed:
revert document change

wenxuanguan commented:
@dongjoon-hyun Thanks for your reply.
I have changed title and description, and reverted the document change.
Thanks for your suggestion, and it really helps me.

HyukjinKwon commented:
ok to test

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10745/
Test PASSed.

SparkQA commented:
**[Test build #105484 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105484/testReport)** for PR 24631 at commit [`2620864`](https://github.com/apache/spark/commit/26208646fd9438cfea5fb36a1d9c60cbf6ed00d8).

SparkQA commented:
**[Test build #105484 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105484/testReport)** for PR 24631 at commit [`2620864`](https://github.com/apache/spark/commit/26208646fd9438cfea5fb36a1d9c60cbf6ed00d8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105484/
Test PASSed.

srowen commented:
This is probably OK but are there more instances like this?

HyukjinKwon reviewed (approved):
Yes, looks fine but it would be better if we fix them more

wenxuanguan committed:
avoid hardcoded config in SparkSubmit

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10753/
Test PASSed.

SparkQA commented:
**[Test build #105493 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105493/testReport)** for PR 24631 at commit [`8069d8b`](https://github.com/apache/spark/commit/8069d8b3eac3777186bf1d5f3d02007543fc69f5).

wenxuanguan commented:
> This is probably OK but are there more instances like this?

I found nothing so far except `SparkSubmit`, and I have fixed it.

SparkQA commented:
**[Test build #105493 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105493/testReport)** for PR 24631 at commit [`8069d8b`](https://github.com/apache/spark/commit/8069d8b3eac3777186bf1d5f3d02007543fc69f5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105493/
Test PASSed.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/SparkConf.scala:
```diff
@@ -595,7 +595,7 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging with Seria
     require(executorTimeoutThresholdMs > executorHeartbeatIntervalMs, "The value of " +
       s"${networkTimeout}=${executorTimeoutThresholdMs}ms must be no less than the value of " +
-      s"spark.executor.heartbeatInterval=${executorHeartbeatIntervalMs}ms.")
+      s"${EXECUTOR_HEARTBEAT_INTERVAL.key}=${executorHeartbeatIntervalMs}ms.")
```
Could you update `SparkContextSuite.scala`, too?
```
-      .set("spark.executor.heartbeatInterval", "1s")
+      .set(EXECUTOR_HEARTBEAT_INTERVAL.key, "1s")
```

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/SparkConf.scala:
```diff
@@ -667,12 +667,12 @@ private[spark] object SparkConf extends Logging {
         translation = s => s"${s.toLong * 10}s")),
     REDUCER_MAX_SIZE_IN_FLIGHT.key -> Seq(
       AlternateConfig("spark.reducer.maxMbInFlight", "1.4")),
-    "spark.kryoserializer.buffer" -> Seq(
```
For these `spark.kryoserializer.buffer` and `spark.kryoserializer.buffer.max`, it seems there are too many instances.
```
core/src/test/scala/org/apache/spark/SparkConfSuite.scala:    assert(conf.getSizeAsKb("spark.kryoserializer.buffer") === 1100)
core/src/test/scala/org/apache/spark/storage/BlockManagerReplicationSuite.scala:  conf.set("spark.kryoserializer.buffer", "1m")
core/src/test/scala/org/apache/spark/storage/BlockManagerReplicationSuite.scala:  conf.set("spark.kryoserializer.buffer", "1m")
core/src/test/scala/org/apache/spark/storage/BlockManagerReplicationSuite.scala:  conf.set("spark.kryoserializer.buffer", "1m")
core/src/test/scala/org/apache/spark/storage/BlockManagerSuite.scala:  val serializer = new KryoSerializer(new SparkConf(false).set("spark.kryoserializer.buffer", "1m"))
core/src/test/scala/org/apache/spark/storage/BlockManagerSuite.scala:      .set("spark.kryoserializer.buffer", "1m")
core/src/test/scala/org/apache/spark/storage/MemoryStoreSuite.scala:  val serializer = new KryoSerializer(new SparkConf(false).set("spark.kryoserializer.buffer", "1m"))
examples/src/main/scala/org/apache/spark/examples/mllib/MovieLensALS.scala:        .set("spark.kryoserializer.buffer", "8m")
```

```
mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:        sc.conf.get("spark.kryoserializer.buffer.max", "64m"))
mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala:        spark.conf.get("spark.kryoserializer.buffer.max", "64m"))
```

dongjoon-hyun replied:
You can use `git grep` to search these instances, @wenxuanguan .

wenxuanguan replied:
@dongjoon-hyun Sorry for delay. 
I have fix the test, and changed some configs with the same problem.

wenxuanguan committed:
fix hardcoded config in test

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10795/
Test PASSed.

SparkQA commented:
**[Test build #105539 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105539/testReport)** for PR 24631 at commit [`b54c18f`](https://github.com/apache/spark/commit/b54c18fbed2bf28c88e2fff0966af00cc9b1b4a3).

SparkQA commented:
**[Test build #105539 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105539/testReport)** for PR 24631 at commit [`b54c18f`](https://github.com/apache/spark/commit/b54c18fbed2bf28c88e2fff0966af00cc9b1b4a3).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105539/
Test FAILed.

wenxuanguan commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10796/
Test PASSed.

SparkQA commented:
**[Test build #105540 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105540/testReport)** for PR 24631 at commit [`ecef09f`](https://github.com/apache/spark/commit/ecef09fe8f901b4e1595faeca715894cbf62126c).

SparkQA commented:
**[Test build #105540 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105540/testReport)** for PR 24631 at commit [`ecef09f`](https://github.com/apache/spark/commit/ecef09fe8f901b4e1595faeca715894cbf62126c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105540/
Test PASSed.

dongjoon-hyun commented:
@wenxuanguan . Thank you for improving the PR. Now, this PR seems need a JIRA issue because it touches many files. Could you file a Apache Spark JIRA issue? Please enumerate the configs in the JIRA which you replaced. Also, please use that JIRA id instead of `[MINOR]`.

wenxuanguan commented:
@dongjoon-hyun Jira SPARK-27774 has been made, and linked with this pr.

srowen commented on examples/src/main/scala/org/apache/spark/examples/mllib/MovieLensALS.scala:
```diff
@@ -24,6 +24,7 @@ import org.apache.log4j.{Level, Logger}
 import scopt.OptionParser
 
 import org.apache.spark.{SparkConf, SparkContext}
+import org.apache.spark.internal.config.Kryo._
```
Hm, I think we might not want to modify the examples actually. They demonstrate what users might run. But they won't be able to import internal classes like this, no?

wenxuanguan replied:
@srowen That make sense, revert it.

wenxuanguan committed:
revert config changes in examples

wenxuanguan commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10840/
Test PASSed.

SparkQA commented:
**[Test build #105586 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105586/testReport)** for PR 24631 at commit [`580c618`](https://github.com/apache/spark/commit/580c618346318b4c73892326adee4c492113793f).

SparkQA commented:
**[Test build #105586 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105586/testReport)** for PR 24631 at commit [`580c618`](https://github.com/apache/spark/commit/580c618346318b4c73892326adee4c492113793f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105586/
Test FAILed.

wenxuanguan commented:
retest this please

SparkQA commented:
**[Test build #105610 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105610/testReport)** for PR 24631 at commit [`d026b8f`](https://github.com/apache/spark/commit/d026b8f15713b2dd6fe29c4f4a201084429c7928).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10862/
Test PASSed.

SparkQA commented:
**[Test build #105610 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105610/testReport)** for PR 24631 at commit [`d026b8f`](https://github.com/apache/spark/commit/d026b8f15713b2dd6fe29c4f4a201084429c7928).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105610/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10875/
Test PASSed.

wenxuanguan commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10878/
Test PASSed.

SparkQA commented:
**[Test build #105625 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105625/testReport)** for PR 24631 at commit [`3337183`](https://github.com/apache/spark/commit/3337183a1059ca0a576696eede985d75ead8dd70).

SparkQA commented:
**[Test build #105625 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105625/testReport)** for PR 24631 at commit [`3337183`](https://github.com/apache/spark/commit/3337183a1059ca0a576696eede985d75ead8dd70).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105625/
Test PASSed.

HyukjinKwon commented:
Merged to master.

wenxuanguan commented:
@HyukjinKwon @srowen @dongjoon-hyun  Thank you for review
